### DISCUSSION TITLE
Is there a difference between a TypeVar and Any if it is unused.

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#8257] [Q&A] [answered]
Specifically in a case like this

```
def func(x: list[Any]) -> bool:
    return bool(x)
```

vs

```
def func(x: list[T]) -> bool:
    return bool(x)
```

Is there functionally a difference and what would be considered a better practice if there isnt.

**Comment by erictraut** [comment]
In general, `Any` has a _very_ different meaning from a type variable. In your simplistic example, the two effectively result in the same behavior, but in slightly more complex examples, the behavior differs. For example:

```python
def func(x: list[Any]):
    y = x[0] # y has type Any
    y.do_something() # No type error
```

```python
def func[T](x: list[T]):
    y = x[0] # y has type T
    y.do_something() # Type error
```

I would not recommend using a type variable even in your simple example. A type variable should appear at least twice in a function's signature — typically in both the input signature and the return type. If you find yourself using a type variable in just one place within a signature, that's a sign that you're using a type variable inappropriately or (at best) unnecessarily.

**Reply by JanEricNitschke** [reply]
Thanks. The point with the `Any` definitely makes sense.

If you wouldnt use the type variable in the case i posted, then how would you do that when using pyright strict mode? Because if i dont specify anything there, then i get a partially unknown warning.

The concrete use case is writing type stubs for pandas that handle the issue described in the [previous discussion](https://github.com/microsoft/pyright/discussions/8250). I guess from a pure stub perspective there wouldnt be a functional difference between `Any` and `T` but ill probably move to a generic TypeVar since that would definitely be more correct if moved over to pandas directly at some point.

**Reply by erictraut** [reply]
You won't get a "partially unknown" warning if you explicitly use `Any` in the type annotation. The `Unknown` type represents an "implicit `Any`" and appears if you omit the type arguments from `list` (or, in the previous discussion, `Series` and `Index`).

**Reply by JanEricNitschke** [reply]
Right. So what do you think is best to use in such a simple case as above or even specifically for the pandas-stubs [here](https://github.com/pandas-dev/pandas-stubs/pull/945).

`Any` or a (unbounded) TypeVar?

**Reply by erictraut** [reply, marked as answer]
Definitely use `Any`, not a `TypeVar`. A `TypeVar` is unnecessary and confusing at best (and inappropriate and incorrect at worst). TypeVars are also way more expensive for static type checkers to handle because they must gather constraints and then apply those constraints to "solve" the type variable. With `Any`, there is no such work required.

**Reply by JanEricNitschke** [reply]
Awesome, thanks again for the help! Really appreciate you taking the time and helping/explaining these things!
